compiler/rustc_codegen_rmc/src/codegen/function.rs

```diff
@@ -158,30 +148,49 @@ impl<'tcx> GotocCtx<'tcx> {
             ),
         };
 
-        // Now that we have the tuple, write the individual component locals
-        // back to it as a GotoC struct.
+        // When we codegen the function signature elsewhere, we will codegen the untupled version.
+        // We then marshall the arguments into a local variable holding the expected tuple.
+        // For a function with args f(a: t1, b: t2, c: t3), the tuple type will look like
+        // ```
+        //    struct T {
+        //        0: t1,
+        //        1: t2,
+        //        2: t3,
+        // }
+        // ```
+        let tup_typ = self.codegen_ty(self.monomorphize(spread_data.ty));
+
+        // We need to marshall the arguments into the tuple
         let tupe = sig.inputs().last().unwrap();
         let args: Vec<&TyS<'tcx>> = match tupe.kind() {
             ty::Tuple(substs) => substs.iter().map(|s| s.expect_ty()).collect(),
             _ => unreachable!("a function's spread argument must be a tuple"),
         };
-
-        // Convert each arg to a GotoC expression.
-        let mut arg_exprs = Vec::new();
         let starting_idx = sig.inputs().len();
-        for (arg_i, arg_t) in args.iter().enumerate() {
-            // The components come at the end, so offset by the untupled length.
-            let lc = Local::from_usize(arg_i + starting_idx);
-            let (name, base_name) = self.codegen_spread_arg_name(&lc);
-            let sym = Symbol::variable(name, base_name, self.codegen_ty(arg_t), loc.clone());
-            self.symbol_table.insert(sym.clone());
-            arg_exprs.push(sym.to_expr());
-        }
-
-        // Finally, combine the expression into a struct.
-        let tuple_expr = Expr::struct_expr_from_values(tup_typ, arg_exprs, &self.symbol_table)
-            .with_location(loc.clone());
-        self.current_fn_mut().push_onto_block(Stmt::decl(tup_sym.to_expr(), Some(tuple_expr), loc));
+        let marshalled_tuple_fields =
+            BTreeMap::from_iter(args.iter().enumerate().map(|(arg_i, arg_t)| {
+                // The components come at the end, so offset by the untupled length.
+                let lc = Local::from_usize(arg_i + starting_idx);
+                let (name, base_name) = self.codegen_spread_arg_name(&lc);
+                let sym = Symbol::variable(name, base_name, self.codegen_ty(arg_t), loc.clone());
+                // TODO, I'm concerned that these are never declared inside the function?
```

Comment:
Have you tried --gen-c to see what's being generated?

Reply:
Added comments based on the results of `--gen-c`